docs/references/architecture/adr-112-txi.md

```diff
@@ -0,0 +1,84 @@
+# ADR 112: Modular transaction hashing
+
+## Changelog
+
+- 2024-02-05: First version (@melekes)
+
+## Status
+
+TBD
+
+## Context
+
+Transaction hashing in CometBFT is currently implemented using `tmhash`
+package, which itself relies on `sha256` to calculate a transaction's hash.
+
+The hash is then used by the built-in indexer (to index this
+transaction) and by the RPC `tx` and `tx_search` endpoints, which allow users
+to search for a transaction using a hash.
+
+`tmhash` is also used to calculate various hashes in CometBFT (e.g.,
+evidence, consensus params, commit, header, partset header).
+
+The problem some application developers are facing is a mismatch between the
+internal/app representation of transactions and the one employed by CometBFT. For
+example, [Evmos](https://evmos.org/) wants transactions to be hashed using
+the [RLP][rlp].
+
+In order to be flexible, CometBFT needs to allow changing the hashing algorithm
+if desired by the app developers.
+
+## Alternative Approaches
+
+None.
+
+## Decision
+
+Give app developers a way to provide their own hash function.
+
+## Detailed Design
+
+Add `hashFn HashFn` argument to `NewNode` constructor in `node.go`.
+
+```go
+type HashFn interface {
+    // New returns a new hash.Hash calculating the given hash function.
+    New() hash.Hash
+    // Size returns the length, in bytes, of a digest resulting from the given hash function.
+    Size() int
+}
+
+type TMHash struct {}
+func (TMHash) New() hash.Hash {
+    return tmhash.New()
+}
+func (TMHash) Size(bz []byte) {
+    return tmhash.Size
+}
+
+func DefaultHashFn() HashFn {
+    return TMHash{}
+}
+```
+
+And then use it to calculate a transaction's hash.
+
+## Consequences
+
+### Positive
+
+- Modular transaction hashing
+- Paving a way for the pluggable cryptography. hashFn can later be used to
+  calculate header's hash, evidence hash.
+
+### Negative
+
+- One more argument in `NewNode`
+
+### Neutral
```

Comment:
Not sure if performance should be considered here at some of the consequences? Probably neutral, since it really depends on the hash method adopted.